icu4c/source/tools/genrb/genrb.cpp

```diff
@@ -631,49 +630,71 @@ processFile(const char *filename, const char *cp,
             if( (filename[0] != U_FILE_SEP_CHAR) && (inputDir[dirlen-1] !='.')){
-                uprv_strcpy(openFileName, inputDir);
-                openFileName[dirlen]     = U_FILE_SEP_CHAR;
+                openFileName.append(inputDir, status);
             }
-            openFileName[dirlen + 1] = '\0';
         } else {
-            openFileName = (char *) uprv_malloc(dirlen + filelen + 1);
-
-            /* test for NULL */
-            if(openFileName == NULL) {
-                status = U_MEMORY_ALLOCATION_ERROR;
-                goto finish;
-            }
-
-            uprv_strcpy(openFileName, inputDir);
-
+            openFileName.append(inputDir, status);
         }
     }
+    openFileName.appendPathPart(filename, status);
 
-    uprv_strcat(openFileName, filename);
+    // Test for CharString failure
+    if (U_FAILURE(status)) {
+        return;
+    }
 
-    ucbuf = ucbuf_open(openFileName, &cp,getShowWarning(),TRUE, &status);
+    ucbuf.adoptInstead(ucbuf_open(openFileName.data(), &cp,getShowWarning(),TRUE, &status));
     if(status == U_FILE_ACCESS_ERROR) {
 
-        fprintf(stderr, "couldn't open file %s\n", openFileName == NULL ? filename : openFileName);
-        goto finish;
+        fprintf(stderr, "couldn't open file %s\n", openFileName.data());
+        return;
     }
-    if (ucbuf == NULL || U_FAILURE(status)) {
+    if (ucbuf.isNull() || U_FAILURE(status)) {
         fprintf(stderr, "An error occurred processing file %s. Error: %s\n",
-                openFileName == NULL ? filename : openFileName, u_errorName(status));
-        goto finish;
+                openFileName.data(), u_errorName(status));
+        return;
     }
     /* auto detected popular encodings? */
     if (cp!=NULL && isVerbose()) {
         printf("autodetected encoding %s\n", cp);
     }
     /* Parse the data into an SRBRoot */
-    data.adoptInstead(parse(ucbuf, inputDir, outputDir, filename,
+    data.adoptInstead(parse(ucbuf.getAlias(), inputDir, outputDir, filename,
             !omitBinaryCollation, options[NO_COLLATION_RULES].doesOccur, &status));
 
     if (data.isNull() || U_FAILURE(status)) {
         fprintf(stderr, "couldn't parse the file %s. Error:%s\n", filename, u_errorName(status));
-        goto finish;
+        return;
     }
+
+    // Run filtering before writing pool bundle
+    if (filterDir != nullptr) {
+        CharString filterFileName;
+        filterFileName.append(filterDir, status);
+        filterFileName.appendPathPart(filename, status);
+
+        // Open the file and read it into filter
+        SimpleRuleBasedPathFilter filter;
+        std::ifstream f(filterFileName.data());
+        if (f.fail()) {
+            std::cerr << "genrb error: unable to open " << filterFileName.data() << std::endl;
+            status = U_FILE_ACCESS_ERROR;
+            return;
+        }
+        std::string currentLine;
+        while (std::getline(f, currentLine)) {
+            filter.addRule(currentLine, status);
```

Comment:
Please handle comment lines and empty lines like in other tools (e.g., gennorm2).

Reply:
Done